### PR TITLE
City state resources

### PR DESCRIPTION
**Pull request by Interdice**
This is basically my old pull request but better!

### CODE REVIEW

**Comment by yairm210** [comment]
Was this a feature in Civ V?

**Comment by logicminimal** [comment]
> Was this a feature in Civ V?

absolutely

**Comment by SomeTroglodyte** [comment]
- I remember a Civ5 screenshot somewhere - find and link here?
- `getImprovableTiles` is more specific than its name suggests, placement in CityInfo invites reuse -> `/**` Kdoc and/or more specific name?
- spurious `// it.getTileResource().improvement`
- two blank lines before the new option, none after
- looping over all improvements to find one with a known name when an indexing operation would do
- translation templates missing (I count three)
- amount should be parameterizable at least for translation (is it really gamespeed/era independent?)
- `${improvableTile.getTileResource()}` relies on `NamedStats.toString()` - perfectly OK but I would probably have avoided that, just personal taste..
- If I were to lint your code to my taste I'd remove the `{}` you added for that single-statement `if` but add blanks before the opening `{` of other `if`s

Other than such nitpicking, however, doesn't look bad at all.

**Comment by Interdice** [comment]
@SomeTroglodyte Fixed some of the things that you suggested

**Comment by SomeTroglodyte** [comment]
Nice to see such a speedy reaction, but "Lux-e-ry"? 

Sorry I wasn't clearer about the Kdoc thing. See "https://kotlinlang.org/docs/kotlin-doc.html" - those are used by Android Studio quite nicely. I assume you're using AS... If you haven't done a Kdoc or similar before I suggest:
- Go to the call in the diplomacy sources, hoves mouse over the function name, observe
- Go to the implementation (e.g. select, Ctrl-B) and type `/** Hi there */` into a line just before the function
- Go back to the call, hover mouse, observe.

Those can really help future coders who might be tempted to use that function...

**Comment by Interdice** [comment]
fixed that dumb typo

**Comment by Interdice** [comment]
Will read about Kdoc

**Comment by SomeTroglodyte** [comment]
Just remember you got nudged in that direction from someone who managed to _overdo_ their Kdoc's. Also - there's tricks and quirks. Maybe I kept records of some stuff I learned in my [notes](https://github.com/SomeTroglodyte/Scratchpad/tree/master/Unciv-notes), or maybe I forgot.

**Comment by Interdice** [comment]
@SomeTroglodyte 

I made the function more generic and moved the more specific stuff to the declaration of the list.

Thanks for the feedback.

**Comment by SomeTroglodyte** [comment]
`<facepalm/>` That Civ5 screenshot that could show Yair this feature was indeed in Civ5 - I searched in vain in issues - and it's of course in the predecessor PR, and of course thanks to ravignir:
https://github.com/yairm210/Unciv/pull/4752#issuecomment-892394059
image

**Comment by SomeTroglodyte** [comment]
Forgot one nitpick: isn't `.map {it}` a no-op? As always, just food for thought.

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt [hunk 1]
```diff
@@ -258,6 +267,44 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
...
+
+        val improvableTiles = otherCiv.getCapital().getImprovableTiles().filterNot {it.getTileResource().resourceType == ResourceType.Bonus}.toList()
+        val tileImprovements = otherCiv.gameInfo.ruleSet.tileImprovements.filter { it.value.turnsToBuild != 0 }
+        var anyImprovements = false
+
+        for (improvableTile in improvableTiles){
+            for (tileImprovement in tileImprovements.values){
+                if (improvableTile.canBuildImprovement(tileImprovement, otherCiv) && improvableTile.getTileResource().improvement == tileImprovement.name){
+                    anyImprovements = true
+                    val improveTileButton = "Build $tileImprovement on ${improvableTile.getTileResource()} (200 Gold)".toTextButton()
```
Comment (yairm210) [inline core/src/com/unciv/ui/trade/DiplomacyScreen.kt:282]: This should be with templates - "build [improvement] on [resource] ([statCost])" - for translation

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt [hunk 2]
```diff
@@ -174,7 +174,16 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
         }
         diplomacyTable.add(giveGiftButton).row()
         if (isNotPlayersTurn()) giveGiftButton.disable()
-        
+
+
+        val improveTileButton = "Gift Improvement".toTextButton()
```
Comment (yairm210) [inline core/src/com/unciv/ui/trade/DiplomacyScreen.kt:179]: This should only appear in the first place if there are improvable tiles - meaning the function 'check if improvable tiles' should include the bonus resource filter

**Comment by Interdice** [comment]
@yairm210 gotcha will fix tomorrow morning

**Comment by Interdice** [comment]
in the tileimprovement Hashmap, the keys and the values are the same

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt (hunk 2 above)
Comment (Interdice) [reply core/src/com/unciv/ui/trade/DiplomacyScreen.kt:179]: fixed

**Comment by Interdice** [comment]
not sure how to use templates though

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt (hunk 1 above)
Comment (Interdice) [reply core/src/com/unciv/ui/trade/DiplomacyScreen.kt]: did i do the templates correctly? btw the 200 gold is constant

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt [hunk 3]
```diff
@@ -174,7 +174,32 @@ class DiplomacyScreen(val viewingCiv:CivilizationInfo):CameraStageBaseScreen() {
...
+        improveTileButton.onClick {
+            rightSideTable.clear()
+            rightSideTable.add(ScrollPane(getImprovementGiftTable(otherCiv)))
+        }
+
+        otherCivDiplomacyManager.influence = 60f
+        viewingCiv.addGold(999999)
```
Comment (yairm210) [inline core/src/com/unciv/ui/trade/DiplomacyScreen.kt:195]: ??????

**Review by yairm210 — CHANGES_REQUESTED** [review]

**Comment by yairm210** [comment]
For the translation, we need 

Build [improvementName] on [resourceName] (200 Gold) = 
Gift Improvement = 

to be added to the template.properties file

**Comment by logicminimal** [comment]
> For the translation, we need 
> 
> Build [improvementName] on [resourceName] (200 Gold) = 
> Gift Improvement = 
> 
> to be added to the template.properties file

what about just adding [amount] Gold?

File: core/src/com/unciv/ui/trade/DiplomacyScreen.kt (hunk 3 above)
Comment (Interdice) [reply core/src/com/unciv/ui/trade/DiplomacyScreen.kt:195]: sorry debugging code

**Comment by Interdice** [comment]
@yairm210 i did it

File: core/src/com/unciv/logic/city/CityInfo.kt [hunk 4]
```diff
@@ -645,6 +646,8 @@ class CityInfo {
             .map { it.getOwner()?.civName }.filterNotNull().toSet() 
             .distinct().toList()
     }
+    fun getImprovableTiles(): Sequence<TileInfo> = getTiles().map {it}
```
[the remark this answers is not in the document]
Comment (Interdice) [reply core/src/com/unciv/logic/city/CityInfo.kt:649]: done

**Comment by yairm210** [comment]
looks good!
You have conflicts, resolve those and we'll merge :)

**Comment by Interdice** [comment]
yep fixed the conflicts. Ready to merge. :)

**Comment by yairm210** [comment]
The tests show that your translation lines don't end with a space
Add and the tests should pass :)

**Comment by Interdice** [comment]
Fixed!

**Comment by SomeTroglodyte** [comment]
Congrats on your second merged PR!

The short discussion on template.properties has prompted me to add an extra chapter to [the wiki entry on translations](https://github.com/yairm210/Unciv/wiki/Translating#rules-for-templates-added-manually) - please tell me if I explained badly or where I missed a detail.

**Comment by Interdice** [comment]
@SomeTroglodyte Thanks! The documentation looks good

**Pull request merged**